File: source/ch/cyberduck/core/editor/WatchEditor.java
package ch.cyberduck.core.editor;

/*
 * Copyright (c) 2012 David Kocher. All rights reserved.
 * http://cyberduck.ch/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * dkocher@cyberduck.ch
 */

import ch.cyberduck.core.Local;
import ch.cyberduck.core.Path;
import ch.cyberduck.core.Session;
import ch.cyberduck.core.local.Application;
import ch.cyberduck.core.local.ApplicationLauncherFactory;
import ch.cyberduck.core.local.FileWatcher;
import ch.cyberduck.core.local.FileWatcherListener;
import ch.cyberduck.ui.Controller;

import org.apache.log4j.Logger;

import java.io.IOException;

/**
 * An editor listing for file system notifications on a particular folder
 *
 * @version $Id$
 */
public class WatchEditor extends BrowserBackgroundEditor implements FileWatcherListener {
    private static final Logger log = Logger.getLogger(WatchEditor.class);

    private FileWatcher monitor = new FileWatcher();

    /**
     * With custom editor for file type.
     *
     * @param c           Browser
     * @param application Editor application
     * @param path        Remote file
     */
    public WatchEditor(final Controller c, final Session session, final Application application, final Path path) {
        super(c, session, application, path);
    }

    /**
     * Edit and watch the file for changes
     */
    @Override
    public void edit() throws IOException {
        final Application application = this.getApplication();
        if(ApplicationLauncherFactory.get().open(edited.getLocal(), application)) {
            this.watch();
        }
        else {
            throw new IOException(String.format("Failed to open application %s", application.getName()));
        }
    }

    /**
     * Watch the file for changes
     */
    public void watch() throws IOException {
        try {
            monitor.register(edited.getLocal()).await();
        }
        catch(InterruptedException e) {
            throw new IOException(String.format("Failure monitoring file %s", edited.getLocal()), e);
        }
        monitor.addListener(this);
    }

    @Override
    protected void delete() {
        monitor.close(edited.getLocal());
        monitor.removeListener(this);
        super.delete();
    }

    @Override
    public void fileWritten(Local file) {
        if(log.isInfoEnabled()) {
            log.info(String.format("File %s written", file));
        }
        this.save();
    }

    @Override
    public void fileRenamed(Local file) {
        if(log.isInfoEnabled()) {
            log.info(String.format("File %s renamed", file));
        }
        monitor.close(edited.getLocal());
        monitor.removeListener(this);
    }

    @Override
    public void fileDeleted(Local file) {
        if(log.isInfoEnabled()) {
            log.info(String.format("File %s deleted", file));
        }
        monitor.close(edited.getLocal());
        monitor.removeListener(this);
    }

    @Override
    public void fileCreated(Local file) {
        if(log.isInfoEnabled()) {
            log.info(String.format("File %s created", file));
        }
        this.save();
    }
}

File: source/ch/cyberduck/core/editor/WatchEditor.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using System;
using System.Diagnostics;
using System.IO;
using Ch.Cyberduck.Ui.Controller;
using ch.cyberduck.core.local;
using org.apache.log4j;
using Path = ch.cyberduck.core.Path;

namespace Ch.Cyberduck.Core.Editor
{
    public class WatchEditor : Editor
    {
        private static readonly Logger Log = Logger.getLogger(typeof (WatchEditor).FullName);
        private FileSystemWatcher _watcher;

        public WatchEditor(BrowserController controller, Application application, Path path)
            : base(controller, application, path)
        {
        }

        protected override void edit()
        {
            Path path = getEdited();
            Application application = getApplication();

            Process process = new Process();
            if (application == null || Utils.IsBlank(application.getIdentifier()))
            {
                application = ch.cyberduck.core.editor.EditorFactory.instance().getDefaultEditor();
                if (application != null && Utils.IsNotBlank(application.getIdentifier()))
                {
                    process.StartInfo.FileName = application.getIdentifier();
                    process.StartInfo.Arguments = "\"" + getEdited().getLocal().getAbsolute() + "\"";
                }
                else
                {
                    process.StartInfo.FileName = getEdited().getLocal().getAbsolute();
                }
            }
            else
            {
                process.StartInfo.FileName = application.getIdentifier();
                process.StartInfo.Arguments = "\"" + getEdited().getLocal().getAbsolute() + "\"";
            }
            if (Utils.StartProcess(process))
            {
                Watch();
            }
        }

        private void Watch()
        {
            _watcher = new FileSystemWatcher();
            _watcher.Path = getEdited().getLocal().getParent().getAbsolute();
            _watcher.Filter = getEdited().getLocal().getName();
            _watcher.NotifyFilter = NotifyFilters.LastAccess | NotifyFilters.LastWrite
                                    | NotifyFilters.FileName | NotifyFilters.DirectoryName;
            RegisterHandlers();
            // Begin watching.
            _watcher.EnableRaisingEvents = true;
        }

        private void HasRenamed(object sender, RenamedEventArgs e)
        {
            Log.debug(String.Format("HasRenamed: from {0} to {1}", e.OldFullPath, e.FullPath));
            save();
        }

        private void HasChanged(object sender, FileSystemEventArgs e)
        {
            Log.debug("HasChanged:" + e.FullPath);
            save();
        }

        protected override void delete()
        {
            _watcher.EnableRaisingEvents = false;
            RemoveHandlers();
            base.delete();
        }

        private void RegisterHandlers()
        {
            _watcher.Changed += HasChanged;
            _watcher.Renamed += HasRenamed;
        }

        private void RemoveHandlers()
        {
            _watcher.Changed -= HasChanged;
            _watcher.Renamed -= HasRenamed;
        }
    }
}